h2o-automl/src/main/java/ai/h2o/automl/TrainingStepsRegistry.java

```diff
@@ -0,0 +1,106 @@
+import water.Iced;
+import water.nbhm.NonBlockingHashMap;
+import water.util.ArrayUtils;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * The registry responsible for loading all {@link TrainingStepsProvider} using service discovery,
+ * and providing the list of {@link TrainingStep} to execute.
+ */
+public class TrainingStepsRegistry extends Iced<TrainingStepsRegistry> {
+
+    static final NonBlockingHashMap<String, Class<TrainingSteps>> stepsByName = new NonBlockingHashMap<>();
+
+    static {
+        ServiceLoader<TrainingStepsProvider> trainingStepsProviders = ServiceLoader.load(TrainingStepsProvider.class);
+        for (TrainingStepsProvider provider : trainingStepsProviders) {
+            stepsByName.put(provider.getName(), provider.getStepsClass());
+        }
+    }
+
+    private StepDefinition[] _defaultTrainingPlan;
+
+    public TrainingStepsRegistry(StepDefinition[] defaultTrainingPlan) {
+        _defaultTrainingPlan = defaultTrainingPlan;
+    }
+
+    /**
+     * @param aml the AutoML instance responsible to execute the {@link TrainingStep}s.
+     * @return the list of {@link TrainingStep}s to execute according to the default training plan.
+     */
+    public TrainingStep[] getOrderedSteps(AutoML aml) {
+        return getOrderedSteps(aml, _defaultTrainingPlan);
+    }
+
+
+    /**
+     * @param aml the AutoML instance responsible to execute the {@link TrainingStep}s.
+     * @return the list of {@link TrainingStep}s to execute according to the given training plan.
+     */
+    public TrainingStep[] getOrderedSteps(AutoML aml, StepDefinition[] trainingPlan) {
+        aml.eventLog().info(Stage.Workflow, "Loading execution steps: "+Arrays.toString(trainingPlan));
+        List<TrainingStep> orderedSteps = new ArrayList<>();
+        for (StepDefinition def : trainingPlan) {
+            Class<TrainingSteps> clazz = stepsByName.get(def._name);
+            if (clazz == null) {
+                throw new IllegalArgumentException("Missing provider for training steps '"+def._name+"'");
+            }
+            try {
+                TrainingSteps steps = clazz.getConstructor(AutoML.class).newInstance(aml);
+                TrainingStep[] toAdd;
+                if (def._alias != null) {
+                    toAdd = steps.getSteps(def._alias);
+                } else if (def._steps != null) {
+                    toAdd = steps.getSteps(def._steps);
+                    if (toAdd.length < def._steps.length) {
+                        List<String> toAddIds = Stream.of(toAdd).map(s -> s._id).collect(Collectors.toList());
+                        Stream.of(def._steps)
+                                .filter(s -> !toAddIds.contains(s._id))
+                                .forEach(s -> aml.eventLog().warn(Stage.Workflow,
```

Comment:
a warning is generated but where is the step actually skipped?

Reply:
it's done when listing the steps: didn't want to introduce the warning there in `getSteps` as I can't generate a meaningful message there: in general, I prefer generating user-facing messages at a rather high-level.